### PR TITLE
Qnamaker

### PR DESCRIPTION
**Pull request by munozemilio**

### CODE REVIEW

File: packages/qnamaker/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,73 @@
+{
+    // Use IntelliSense to learn about possible attributes.
```
Comment (jspruance) [inline packages/qnamaker/.vscode/launch.json]: typically you can't add comments to a json file (that's why this is highlighted, as a warning).

File: packages/qnamaker/src/commands/qnamaker/create/kb.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    let result = await qnamaker(input.config, input.serviceManifest, flags, input.requestBody)
+
+    if (result.error) {
+      throw new CLIError(JSON.stringify(result.error, null, 4))
+    }
+
+    if (flags.wait) {
+      result = await qnaconfig.waitForOperationSucceeded(input.config, result)
+
+      let kbId = result.resourceLocation.split('/')[2]
```
Comment (jspruance) [inline packages/qnamaker/src/commands/qnamaker/create/kb.ts]: not critical - but better to use 'const' if this is not reassigned

File: packages/qnamaker/src/commands/qnamaker/delete/kb.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      let kbResult = await new Knowledgebase().getKnowledgebaseDetails(input.config)
+      let kb = await JSON.parse(await kbResult.text())
+      let answer = await cli.prompt(`Are you sure you would like to delete ${kb.name} [${kb.id}]? [no] `, {default: 'no'})
+      if (answer.trim()[0] === 'n') {
+        this.log('operation canceled')
+        return
+      }
+    }
+
+    let result = await qnamaker(input.config, input.serviceManifest, flags, input.requestBody)
```
Comment (jspruance) [inline packages/qnamaker/src/commands/qnamaker/delete/kb.ts]: same - should be 'const'

File: packages/qnamaker/utils/api/serviceBase.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+ * Copyright(c) Microsoft Corporation.All rights reserved.
+ * Licensed under the MIT License.
+ */
+const os = require('os');
+const { insertParametersFromObject } = require('../insertParametersFromObject');
+const deriveParamsFromPath = require('../deriveParamsFromPath');
+const packageJSON = require('./../../package');
+
+const fetch = require('node-fetch');
+
```
Comment (jspruance) [inline packages/qnamaker/utils/api/serviceBase.js]: so is this just a wrapper for node-fetch? making it global so you don't have to repeat the wrapper logic everywhere?

File: packages/qnamaker/utils/api/serviceBase.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        let URL = insertParametersFromObject(tokenizedUrl, params);
+        if (method === 'get' && ('skip' in params || 'take' in params)) {
+            const { skip, take } = params;
+            URL += '?';
+            if (!isNaN(+skip)) {
+                URL += `skip=${~~skip}`;
+            }
+            if (!isNaN(+take)) {
+                URL += !isNaN(+skip) ? `&take=${~~take}` : `take=${~~take}`;
```
Comment (jspruance) [inline packages/qnamaker/utils/api/serviceBase.js]: never seen this before (~~) cool!

File: packages/qnamaker/utils/api/serviceBase.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        }
+        const body = dataModel ? JSON.stringify(dataModel) : undefined;
+        if (params.debug) {
+            console.log(`${method.toUpperCase()} ${URL}`);
+            if (headers)
+                console.log(`HEADERS:${JSON.stringify(headers)}`);
+            if (body)
+                console.log(body);
+        }
+        //console.log('url is: '+URL+ " method "+ method)
```
Comment (jspruance) [inline packages/qnamaker/utils/api/serviceBase.js]: console.log

File: packages/qnamaker/utils/kbawaitoperation.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        while (true) {
+            let opResult = await new Operations().getOperationDetails({
+                subscriptionKey: config.subscriptionKey,
+                operationId: result.operationId
+            });
+
+            if (opResult.error)
+                throw new Error(JSON.stringify(opResult.error, null, 4));
+
+            result = await opResult.json();
```
Comment (jspruance) [inline packages/qnamaker/utils/kbawaitoperation.js]: is this a reassignment, or need a var declaration?

File: packages/qnamaker/utils/kbawaitoperation.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+            process.stderr.write(`'\r${result.operationState}`);
+            count++;
+
+            await cli.action.start(result.operationState)
+            for (let x = 0; x < count; x++)
```
Comment (jspruance) [inline packages/qnamaker/utils/kbawaitoperation.js]: use 'map' if possible

File: packages/qnamaker/utils/api/serviceBase.js (hunk 4 above)
Comment (munozemilio) [reply packages/qnamaker/utils/api/serviceBase.js]: thats right

**Pull request merged**